subsys/canbus/canopen/CO_driver.c

```diff
@@ -0,0 +1,488 @@
+{
+	/* No operation */
+}
+
+void CO_CANsetNormalMode(CO_CANmodule_t *CANmodule)
+{
+	CANmodule->CANnormal = true;
+}
+
+CO_ReturnError_t CO_CANmodule_init(CO_CANmodule_t *CANmodule,
+				   s32_t CANbaseAddress,
+				   CO_CANrx_t rxArray[], u16_t rxSize,
+				   CO_CANtx_t txArray[], u16_t txSize,
+				   u16_t CANbitRate)
+{
+	struct device *device;
+	u16_t i;
+	int err;
+
+	LOG_DBG("rxSize = %d, txSize = %d", rxSize, txSize);
+
+	device = canopen_get_can_device(CANbaseAddress);
+
+	if (!CANmodule || !rxArray || !txArray || !device) {
+		LOG_ERR("failed to initialize CAN module %d", CANbaseAddress);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	if (rxSize > CONFIG_CAN_MAX_FILTER) {
+		LOG_ERR("insufficient number of concurrent CAN RX filters"
+			" (needs %d, %d available)", rxSize,
+			CONFIG_CAN_MAX_FILTER);
+		return CO_ERROR_OUT_OF_MEMORY;
+	} else if (rxSize < CONFIG_CAN_MAX_FILTER) {
+		LOG_INF("excessive number of concurrent CAN RX filters enabled"
+			" (needs %d, %d available)", rxSize,
+			CONFIG_CAN_MAX_FILTER);
+	}
+
+	canopen_detach_all_rx_filters(CANmodule);
+	canopen_tx_queue.CANmodule = CANmodule;
+
+	CANmodule->dev = device;
+	CANmodule->rx_array = rxArray;
+	CANmodule->rx_size = rxSize;
+	CANmodule->tx_array = txArray;
+	CANmodule->tx_size = txSize;
+	CANmodule->CANnormal = false;
+	CANmodule->first_tx_msg = true;
+	CANmodule->errors = 0;
+	CANmodule->em = NULL;
+
+	for (i = 0U; i < rxSize; i++) {
+		rxArray[i].ident = 0U;
+		rxArray[i].pFunct = NULL;
+		rxArray[i].filter_id = CAN_NO_FREE_FILTER;
+	}
+
+	for (i = 0U; i < txSize; i++) {
+		txArray[i].bufferFull = false;
+	}
+
+	err = can_configure(CANmodule->dev, CAN_NORMAL_MODE, KHZ(CANbitRate));
+	if (err) {
+		LOG_ERR("failed to configure CAN interface (err %d)", err);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	CANmodule->configured = true;
+
+	return CO_ERROR_NO;
+}
+
+void CO_CANmodule_disable(CO_CANmodule_t *CANmodule)
+{
+	int err;
+
+	if (!CANmodule || !CANmodule->dev) {
+		return;
+	}
+
+	canopen_detach_all_rx_filters(CANmodule);
+
+	err = can_configure(CANmodule->dev, CAN_SILENT_MODE, 0);
+	if (err) {
+		LOG_ERR("failed to disable CAN interface (err %d)", err);
+	}
+}
+
+CO_ReturnError_t CO_CANrxBufferInit(CO_CANmodule_t *CANmodule, u16_t index,
+				u16_t ident, u16_t mask, bool_t rtr,
+				void *object,
+				CO_CANrxBufferCallback_t pFunct)
+{
+	struct zcan_filter filter;
+	CO_CANrx_t *buffer;
+
+	if (!CANmodule || !pFunct || (index >= CANmodule->rx_size)) {
+		LOG_ERR("failed to initialize CAN rx buffer, illegal argument");
+		CO_errorReport(CANmodule->em, CO_EM_GENERIC_SOFTWARE_ERROR,
+			       CO_EMC_SOFTWARE_INTERNAL, 0);
+		return CO_ERROR_ILLEGAL_ARGUMENT;
+	}
+
+	buffer = &CANmodule->rx_array[index];
+	buffer->object = object;
+	buffer->pFunct = pFunct;
+
+	filter.id_type = CAN_STANDARD_IDENTIFIER;
+	filter.std_id = ident;
+	filter.std_id_mask = mask;
+	filter.rtr = (rtr ? 1 : 0);
+	filter.rtr_mask = (rtr ? 1 : 0);
```

Comment:
Don't you want to always enable the mask for RTR?

Reply:
@alexanderwachter Good catch, thank you! Fixed.